File: quick_sort/quickSort.js
/*
*	Implementation of quicksort algorithm
*	Time complexity: O(nlogn)
* 	Space complexity: O(n)
*	n is the size of array
*/
function median (arr, i, j, k) {
	/*
	:param arr: array of elements
	:param i: index of first element
	:param j: index of last element
	:param k: index of middle element
	:return: return median of values at indices i, j and k
	*/
	if (arr[i] > arr[j] && arr[i] > arr[k]) {
		if (arr[k] > arr[j]) {
			return k;
		} else {
			return j;
		}
	} else if (arr[j] > arr[i] && arr[j] > arr[k]) {
		if (arr[k] > arr[i]) {
			return k;
		} else {
			return i;
		}
	} else {
		if (arr[i] > arr[j]) {
			return i;
		} else {
			return j;
		}
	}
}

function partition (arr, start, end) {
	/*
	:param arr: array of elements
	:param start: index of first element
	:param end: index of last element
	:return: return value for function, used in partitioning of array
	*/
	let j = start - 1;
	let tmp;
	let pi = median(arr, start, end, parseInt((start + end) / 2));
	tmp = arr[pi];
	arr[pi] = arr[end];
	arr[end] = tmp;
	let pivot = arr[end];
	for (let i = start; i < end; i++) {
		if (arr[i] <= pivot) {
			j++;
			tmp = arr[i];
			arr[i] = arr[j];
			arr[j] = tmp;
		}
	}
	tmp = arr[j + 1];
	arr[j + 1] = arr[end];
	arr[end] = tmp;
	return j + 1;
}

function quickSort (arr, left = 0, right = arr.length - 1) {
	/*
	:param arr: array to be sorted
	:param left: index of first element
	:param right: index of last element
	*/
	if (left < right) {
		let p = partition(arr, left, right);
		quickSort(arr, left, p - 1);
		quickSort(arr, p + 1, right);
	}
}

function main () {
	let arr = [2, 1, 6, 44, 8, 9, 10];
	quickSort(arr);
	console.log('Sorted data is', arr);
}

main();

File: quick_sort/QuickSort.java
public class QuickSort {

    private static int compare(int[] arr, int i, int j) {
        if (arr[i] > arr[j]) {
            return i;
        } else {
            return j;
        }
    }

    /*
    This method returns the index of median of array at i, j, k.
    @param arr: Input array to be sorted.
    @param i: Left end index of array.
    @param j: Right end index of array.
    @param k: Middle index of array.
    @return: Index of median of array at i, j, k.
    */
    private static int median(int[] arr, int i, int j, int k) {
        if (arr[i] > arr[j] && arr[i] > arr[k]) {
            return compare(arr, k, j);
        } else if (arr[j] > arr[i] && arr[j] > arr[k]) {
            return compare(arr, k, i);
        } else {
            return compare(arr, i, j);
        }
    }

    private static int partition(int[] arr, int start, int end) {
        int l = median(arr, start, end, (start + end) / 2);
        int p_idx = start - 1;
        int tmp = arr[l];
        arr[l] = arr[end];
        arr[end] = tmp;
        int pivot = arr[end];
        for (int i = start; i < end; ++i) {
            if (arr[i] <= pivot) {
                p_idx++;
                tmp = arr[i];
                arr[i] = arr[p_idx];
                arr[p_idx] = tmp;
            }
        }
        p_idx++;
        tmp = arr[p_idx];
        arr[p_idx] = arr[end];
        arr[end] = tmp;
        return p_idx;
    }

    public static void quickSort(int[] a, int left, int right) {
        if (left < right) {
            int pi = partition(a, left, right);  //pi index of pivot
            quickSort(a, left, pi - 1);  //sort left of pivot
            quickSort(a, pi + 1, right);  //sort right of pivot
        }
    }

    public static void main(String[] args) {
        int[] arr = new int[] {2, 4, 2, 6, 7, -1};
        quickSort(arr, 0, arr.length - 1);
        for (int element : arr) {
            System.out.println(element);
        }
    }
}
